### PR TITLE
Add Synchronous instruments

### PR DESCRIPTION
**Pull request by MadVikingGod**
Closes: #2814 

This implements the sync instrument providers.

### CODE REVIEW

File: sdk/metric/instrument.go [hunk 1]
```diff
@@ -53,3 +62,21 @@ func (i *instrumentImpl[N]) Observe(ctx context.Context, val N, attrs ...attribu
...
+	if err := ctx.Err(); err != nil {
+		return
+	}
+	for _, agg := range i.aggregators {
+		agg.Aggregate(val, attribute.NewSet(attrs...))
+	}
+}
+
+func (i *instrumentImpl[N]) Record(ctx context.Context, val N, attrs ...attribute.KeyValue) {
+	if err := ctx.Err(); err != nil {
```
Comment (MrAlias) [inline sdk/metric/instrument.go:76]: These are equivalent, just a method name difference, right?

```suggestion
func (i *instrumentImpl[N]) Add(ctx context.Context, val N, attrs ...attribute.KeyValue) {
	i.aggregate(ctx, val, attrs)
}

func (i *instrumentImpl[N]) Record(ctx context.Context, val N, attrs ...attribute.KeyValue) {
	i.aggregate(ctx, val, attrs)
}

func (i *instrumentImpl[N]) aggregate(ctx context.Context, val N, attrs []attribute.KeyValue) {
	if err := ctx.Err(); err != nil {
		return
	}
	for _, agg := range i.aggregators {
		agg.Aggregate(val, attribute.NewSet(attrs))
	}
}
```

**Review by MrAlias — APPROVED** [review]

File: sdk/metric/instrument.go (hunk 1 above)
Comment (MadVikingGod) [reply sdk/metric/instrument.go]: You are right, it's also the same code in Observe().  Refactored the lot of them.

File: sdk/metric/instrument_provider.go [hunk 2]
```diff
@@ -149,3 +151,125 @@ func (p asyncFloat64Provider) Gauge(name string, opts ...instrument.Option) (asy
...
+	}, cfg.Unit())
+	if len(aggs) == 0 && err != nil {
+		err = fmt.Errorf("instrument does not match any view: %w", err)
+	}
+	return &instrumentImpl[float64]{
+		aggregators: aggs,
+	}, err
+}
+
+// Gauge creates an instrument for recording the current value.
```
Comment (dashpole) [inline sdk/metric/instrument_provider.go:259]: Comment is for gauge.  Should be for histogram instead.

File: sdk/metric/instrument_provider.go [hunk 3]
```diff
@@ -149,3 +151,125 @@ func (p asyncFloat64Provider) Gauge(name string, opts ...instrument.Option) (asy
...
+	}, cfg.Unit())
+	if len(aggs) == 0 && err != nil {
+		err = fmt.Errorf("instrument does not match any view: %w", err)
+	}
+	return &instrumentImpl[int64]{
+		aggregators: aggs,
+	}, err
+}
+
+// Gauge creates an instrument for recording the current value.
```
Comment (dashpole) [inline sdk/metric/instrument_provider.go:198]: Same here (gauge vs histogram)

**Review by dashpole — APPROVED** [review]
lgtm after comments are fixed

**Pull request merged**